Enable Docker container health check via shell command

catalan-adobe opened the pull request:
## Description
In our WDIO projects we use Selenium Docker images for which simple URL checking is not enough to determine readiness of the service (ex. [vvoyer/selenium-standalone](https://github.com/vvo/selenium-standalone/tree/master/docker)).

I propose an extension of dockerOptions.healthCheck to poll check the execution of a command until it returns 0.

The main use case is to execute a shell command to test the `health` status of a Docker container (for images which implement Docker [`HEALTHCHECK`](https://docs.docker.com/engine/reference/builder/#healthcheck)

## Checklist
- [x] Unit test added
- [x] Updated README
- [x] Code style is consistent with the rest of the project

stsvilik commented:
I'll take a look at this tonight

stsvilik commented on README.md:
````diff
@@ -98,6 +99,21 @@ healthCheck: {
+}
+```
+
+
````
Do you mind reducing empty lines to 1?

stsvilik commented on src/utils/childProcess.js:
```diff
@@ -1,18 +1,14 @@
-        const commands = cmd.split(SPACE);
-        const [app, ...args] = commands;
-        const childProcess = spawn(app, args);
+        const childProcess = exec(cmd);
```
👍 Thats definitely much simpler. I don't remember now why I've chosen `spawn` (probably had some issues with running images in foreground), but it works!

catalan-adobe replied:
Working on below remark about user input sanitisation, I had further look in node.js childProcess API and I think you used `spawn` without shell because it's secured OOTB, no need to sanitise anything.
Now I'm concerned with forcing change to use `exec`.

stsvilik commented on src/utils/childProcess.js:
```diff
@@ -31,9 +27,7 @@ export function runProcess(cmd) {
-        const commands = cmd.split(SPACE);
-        const [app, ...args] = commands;
-        const childProcess = spawn(app, args, { stdio: 'ignore' });
+        let childProcess = exec(cmd, { shell: true });
```
According to [spec](https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback) , `shell` flag in `exec` only takes a string, not a boolean like in `spawn`. There are also plenty of warning about using `shell` switch with uninitialized user input. Perhaps add some sanitization to strip out `>`,`&` and other potentially dangerous metacharacters?

stsvilik replied:
I ran tests w/o `shell: true` and they all passed. Does this mean test is not sufficient or that this flag is not necessary?

catalan-adobe replied:
Thanks a lot for this constructive feedback.
* About `shell` flag, good catch! I guess I kept that from one of dev iteration still using [`childProcess.spawn`](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options). I will simply remove it.
* Removing this, do you still recommend `cmd` sanitisation?

stsvilik commented:
@catalan-adobe If the goal is to leverage Docker native `HEALTHCHECK` functionality, then I would definitely defer to that instead of direct CLI query. From your comment I understand that it would still require polling, but this time against `docker inspect`. As you mentioned, `inspect` command returns `health` value when the feature is configured in the `Dockerfile`. I already perform an inspect query in the code at least once to determine if image needs to be pulled or is already present. What you can do is use that call to see if `health` is present and set off a flag which will determine the healthcheck polling mechanics. I am a bit torn between an explicit flag to perform `inspect` poling or an automatic one when `HEALTHCHECK` feature is detected. Perhaps the right approach is to use `dockerOptions.healthCheck.inspect` defaulted to true if native healthcheck is detected.

stsvilik commented:
@catalan-adobe I read a little more on Docker healthcheck and it appears that there is an event emitted by docker when `health_status` changes. This would be better than polling, though wrapping `docker events` with node will require some work.

catalan-adobe commented:
Thanks again for the feedback. I reverted my first commit and implemented a process which automatically checks if the Docker image has a HEALTHCHECK and:
* if yes, automatically poll checks health status via `docker inspect ...` command
* if no, falls back to Ping check
(Advantage is it does not change dockerOptions.healthCheck object API)

Let me know what you think about this version. If implementation is ok for you I would then ask some help/guidance to properly unit test the newly added internal functions.

stsvilik commented on src/utils/docker.js:
```diff
@@ -172,6 +182,52 @@ class Docker extends EventEmitter {
+        return new Promise((resolve, reject) => {
+            runCommand(`docker inspect ${ this.image }`).then(function(output) {
+                try {
+                    let inspectJson = JSON.parse(output);
```
Wouldn't `output` be a `UIntArray`? I don't think `JSON.parse` takes that.

stsvilik commented on src/utils/childProcess.js:
```diff
@@ -33,15 +33,20 @@ export function runCommand(cmd) {
+        const childProcess = spawn(app, args);
+        let stdout = '';
+
+        childProcess.stdout.on('data', (data) => {
```
If you want `data` to be a `String`, you would need to set encoding (i.e `childProcess.stdout.setEncoding('utf-8');`

stsvilik commented:
@catalan-adobe current approach would work, however I started investigating approach of binding to Docker events and managed to put together code which can fire events for practically anything that Docker service does (more info [here](https://docs.docker.com/engine/reference/commandline/events/#filter-events-by-time)). I think if we use those, then we would need to poll :). What are your thoughts?

stsvilik commented:
Yeah, lets re-open once docker events are wired-in, then you can add healthcheck code :)

catalan-adobe commented:
Ok perfect, thanks a lot for taking good care of the PR! :)

stsvilik commented:
FYI, I'm wrapping up tests for new code.

stsvilik commented:
Could you take a look at #42

stsvilik commented:
@catalan-adobe would you mind taking a look at PR #42 ?

catalan-adobe commented:
@stsvilik, I missed the first notification. I had already a look at the PR and as I was not in the reviewers list I did not take the opportunity to comment/approve it.

stsvilik commented:
hi @catalan-adobe - are you still interested in doing this PR?
